kcp-dev/kcp-glbc
# Webhook service and reconciliation

## Description of Changes

Add the logic and configuration for the controller to serve webhook requests, as well as reconcile ValidatingWebhookConfiguration for DomainVerification

## Release and Testing

To verify locally, run the controller as a Deployment in the local kind cluster.

### 1. Run the local setup
```sh
make local-setup
```

### 2. Enable webhooks 

1. Ensure the following env variables are set
    ```
    GLBC_TLS_PROVIDED=true
    GLBC_TLS_PROVIDER=glbc-ca
    GLBC_WEBHOOKS_PORT=8443
    ```

### 3. Deploy the controller
```sh
export KUBECONFIG=.kcp/admin.kubeconfig
make kind-build-load-and-deploy-image
```
This command builds the image, loads it into the kind cluster and deploys it

### 4. Verify that the webhook resources have been created (it might take a while for the controller to create it after deploying it)

1. Verify the `ValidatingWebhookConfiguration`
    ```sh
    kubectl get validatingwebhookconfigurations glbc.domainverification.dev -o yaml
    ```
    ```yaml
    apiVersion: admissionregistration.k8s.io/v1
    kind: ValidatingWebhookConfiguration
    metadata:
      annotations:
        kcp.dev/cluster: root:kuadrant
      creationTimestamp: "2022-09-08T12:31:32Z"
      generation: 3
      name: glbc.domainverification.dev
      resourceVersion: "1084"
      uid: 45570263-756a-4017-b960-58481b4357c4
    webhooks:
    - admissionReviewVersions:
      - v1
      clientConfig:
        caBundle: ...
        url: https://cccu194k9onsd8pvfrmg.dev.hcpapps.net/domainverifications
      failurePolicy: Fail
      matchPolicy: Exact
      name: glbc.domainverification.dev
      namespaceSelector: {}
      objectSelector: {}
      rules:
      - apiGroups:
        - kuadrant.dev
        apiVersions:
        - v1
        operations:
        - CREATE
        - UPDATE
        resources:
        - domainverifications
        scope: '*'
      sideEffects: None
      timeoutSeconds: 5
    ```
3. Edit the `hosts` file to make the controller domain available
    1. Get the auto-generated domain and the IP from the controller Ingress 
        ```sh
        kubectl get ingress ingress-glbc -n kcp-glbc -o yaml
        ```
    2. (If using `fake` DNS) Add it to `/etc/hosts`. Example:
        ```
        172.21.0.2 cccu194k9onsd8pvfrmg.dev.hcpapps.net
        ```

2. Create a `DomainVerification` resource and verify that the controller logs the webhook request
   In the controller logs:
    ```
    INFO webhooks-server admission/webhook.go:112 Got a webhook request
    ```
